pkg/ginutil/writeproblem.go

```diff
@@ -152,3 +152,94 @@ func WriteInvalidBindError(c *gin.Context, err error, detail string) {
+		Detail: detail,
+	})
+}
+
+// WriteAPIWriteError uses WriteProblemError to write a 502 "Bad Gateway"
+// response with the type "/prob/api/unexpected-api-write-error".
+//
+// Meant to be used on unexpected error when writing data using the Wharf API.
+func WriteAPIWriteError(c *gin.Context, err error, detail string) {
+	WriteProblemError(c, err, problem.Response{
+		Type: "/prob/provider/unexpected-api-write-error",
+		Title: "Unexpected API write error.",
+		Status: http.StatusBadGateway,
+		Detail: detail,
+	})
+}
+
+// WriteResponseFormatError uses WriteProblemError to write a 502 "Bad Gateway"
+// response with the type "/prob/provider/unexpected-response-format".
+//
+// Meant to be used on unexpected error when the response format does not match our expectations.
+func WriteResponseFormatError(c *gin.Context, err error, detail string) {
+	WriteProblemError(c, err, problem.Response{
+		Type: "/prob/provider/unexpected-response-format",
+		Title: "Unexpected response format.",
+		Status: http.StatusBadGateway,
+		Detail: detail,
+	})
+}
+
+// WriteFetchBuildDefinitionError uses WriteProblemError to write a 400 "Bad Request"
+// response with the type "/prob/provider/fetch-build-definition".
+//
+// Meant to be used on error when fetching the build definition.
+func WriteFetchBuildDefinitionError(c *gin.Context, err error, detail string) {
+	WriteProblemError(c, err, problem.Response{
+		Type: "/prob/provider/fetch-build-definition",
+		Title: "Error fetching build definition.",
+		Status: http.StatusBadRequest,
+		Detail: detail,
+	})
+}
+
+// WriteComposingProviderDataError uses WriteProblemError to write a 400 "Bad Request"
+// response with the type "/prob/provider/composing-provider-data".
+//
+// Meant to be used on error when composing provider data.
+func WriteComposingProviderDataError(c *gin.Context, err error, detail string) {
+	WriteProblemError(c, err, problem.Response{
+		Type: "/prob/provider/composing-provider-data",
+		Title: "Error composing provider data.",
+		Status: http.StatusBadRequest,
+		Detail: detail,
+	})
+}
+
+// WriteTriggerError uses WriteProblemError to write a 502 "Bad Gateway"
+// response with the type "/prob/build/unexpected-trigger-error".
+//
+// Meant to be used on unexpected error during build process regarding a trigger.
+func WriteTriggerError(c *gin.Context, err error, detail string) {
+	WriteProblemError(c, err, problem.Response{
+		Type: "/prob/build/unexpected-trigger-error",
+		Title: "Unexpected trigger error.",
+		Status: http.StatusBadGateway,
+		Detail: detail,
+	})
+}
+
+// WriteAuthenticationError uses WriteProblemError to write a 401 "Unauthorized"
+// response with the type "prob/provider/authentication-error".
+//
+// Meant to be used for failed authentication.
+func WriteAuthenticationError(c *gin.Context, err error, detail string) {
+	WriteProblemError(c, err, problem.Response{
+		Type: "prob/provider/authentication-error",
+		Title: "Error authenticating.",
```

Comment:
Missing leading slash `/`, and this problem does not seem to be strictly related to the providers.

Maybe just file this under `/prob/api/`, and then we could use this function in the wharf-api repo as well?

And IMO "unauthorized" is clearer. It's still a negative word, so no need to include "Error" in the title or type.

Maybe also accompany this function with a `WriteUnauthorized` without an `error` parameter.

```suggestion
// WriteUnauthorizedError uses WriteProblemError to write a 401 "Unauthorized"
// response with the type "/prob/api/unauthorized".
//
// Meant to be used for failed authentication.
func WriteUnauthorizedError(c *gin.Context, err error, detail string) {
	WriteProblemError(c, err, problem.Response{
		Type: "/prob/api/unauthorized",
		Title: "Unauthorized.",
```

Reply:
By accompany with, do you mean an additional function, like:
```go
// WriteUnauthorized uses WriteProblem to write a 401 "Unauthorized"
// response with the type "/prob/api/unauthorized".
//
// Meant to be used for failed authentication.
func WriteUnauthorized(c *gin.Context, detail string) {
	WriteProblem(c, problem.Response{
		Type: "/prob/api/unauthorized",
		Title: "Unauthorized.",
		Status: http.StatusUnauthorized,
		Detail: detail,
	})
}
```
When would you use one over the other?

Reply:
If there was no `error` to include but still detected the user as unauthorized.